include/forcing/ForcingsEngineDataProvider.hpp

```diff
@@ -0,0 +1,168 @@
+struct ForcingsEngineDataProvider
+  : public DataProvider<DataType, SelectionType>
+{
+    using data_type = DataType;
+    using selection_type = SelectionType;
+    using clock_type = std::chrono::system_clock;
+
+    ~ForcingsEngineDataProvider() = default;
+
+    boost::span<const std::string> get_available_variable_names() override
+    {
+        return var_output_names_;
+    }
+
+    long get_data_start_time() override
+    {
+        return clock_type::to_time_t(time_begin_);
+    }
+  
+    long get_data_stop_time() override
+    {
+        return clock_type::to_time_t(time_end_);
+    }
+
+    long record_duration() override
+    {
+        return std::chrono::duration_cast<std::chrono::seconds>(time_step_).count();
+    }
+
+    size_t get_ts_index_for_time(const time_t& epoch_time) override
+    {
+        const auto epoch = clock_type::from_time_t(epoch_time);
+
+        if (epoch < time_begin_ || epoch > time_end_) {
+          throw std::out_of_range{
+            "epoch " + std::to_string(epoch.time_since_epoch().count()) +
+            " out of range of " + std::to_string(time_begin_.time_since_epoch().count()) + ", "
+            + std::to_string(time_end_.time_since_epoch().count())
+          };
+        }
+
+        return (epoch - time_begin_) / time_step_;
+    }
+
+    // Temporary (?) function to clear out instances of this type.
+    static void finalize_all() {
+        instances_.clear();
+    }
+
+    /* Remaining virtual member functions from DataProvider must be implemented
+       by derived classes. */
+
+    double get_value(const selection_type& selector, data_access::ReSampleMethod m) override = 0;
+    
+    std::vector<double> get_values(const selection_type& selector, data_access::ReSampleMethod m) override = 0;
+
+
+    /* Friend functions */
+    static ForcingsEngineDataProvider* instance(
+        const std::string& init,
+        const std::string& time_begin,
+        const std::string& time_end,
+        const std::string& time_fmt = default_time_format
+    )
+    {
+        auto& inst = instances_.at(init);
+        if (inst != nullptr) {
+            assert(inst->time_begin_.time_since_epoch() == std::chrono::seconds{parse_time(time_begin, time_fmt)});
+            assert(inst->time_end_.time_since_epoch() == std::chrono::seconds{parse_time(time_end, time_fmt)});
+        }
+
+        return inst.get();
+    }
+
+  protected:
+
+    // TODO: It may make more sense to have time_begin_seconds and time_end_seconds coalesced into
+    //       a single argument: `clock_type::duration time_duration`, since the forcings engine
+    //       manages time via a duration rather than time points. !! Need to double check
+    ForcingsEngineDataProvider(
+      const std::string& init,
+      std::size_t time_begin_seconds,
+      std::size_t time_end_seconds
+    )
+      : time_begin_(std::chrono::seconds{time_begin_seconds})
+      , time_end_(std::chrono::seconds{time_end_seconds})
+    {
```

Comment:
```suggestion
    {
        assert_forcings_engine_requirements();
```
I think this would be the right place to re-introduce this call, since it's on relatively low-count construction of the shared instances, not per-catchment.